af/fastpath/fastpath/detector.py

```diff
@@ -0,0 +1,520 @@
+psycopg2.extras.register_default_jsonb(loads=ujson.loads, globally=True)
+psycopg2.extras.register_default_json(loads=ujson.loads, globally=True)
+
+
+def fetch_past_data(conn, start_date):
+    """Fetch past data in large chunks
+    """
+    chunk_size = 20000
+    q = """
+    SELECT
+        coalesce(false) as anomaly,
+        coalesce(false) as confirmed,
+        input,
+        measurement_start_time,
+        probe_cc,
+        scores::text,
+        test_name,
+        tid
+    FROM fastpath
+    WHERE measurement_start_time >= %(start_date)s
+
+    UNION
+
+    SELECT
+        anomaly,
+        confirmed,
+        input,
+        measurement_start_time,
+        probe_cc,
+        coalesce('') as scores,
+        test_name,
+        coalesce('') as tid
+
+    FROM measurement
+    JOIN report ON report.report_no = measurement.report_no
+    JOIN input ON input.input_no = measurement.input_no
+    WHERE measurement_start_time >= %(start_date)s
+
+    ORDER BY measurement_start_time
+    """
+
+    # FIXME
+    q = """
+    SELECT
+        coalesce(false) as anomaly,
+        coalesce(false) as confirmed,
+        input,
+        measurement_start_time,
+        probe_cc,
+        scores::text,
+        test_name,
+        tid
+    FROM fastpath
+    WHERE measurement_start_time >= %(start_date)s
+
+    ORDER BY measurement_start_time
+    """
+    assert start_date
+    p = dict(start_date=str(start_date))
+
+    with conn.cursor(cursor_factory=psycopg2.extras.DictCursor) as cur:
+        cur.execute(q, p)
+        while True:
+            rows = cur.fetchmany(chunk_size)
+            if not rows:
+                break
+            log.info("Fetched msmt chunk of size %d", len(rows))
+            for r in rows:
+                d = dict(r)
+                if d["scores"]:
+                    d["scores"] = ujson.loads(d["scores"])
+
+                yield d
+
+
+def fetch_past_data_selective(conn, start_date, cc, test_name, inp):
+    """Fetch past data in large chunks
+    """
+    chunk_size = 200_000
+    q = """
+    SELECT
+        coalesce(false) as anomaly,
+        coalesce(false) as confirmed,
+        input,
+        measurement_start_time,
+        probe_cc,
+        scores::text,
+        test_name,
+        tid
+    FROM fastpath
+    WHERE measurement_start_time >= %(start_date)s
+    AND probe_cc = %(cc)s
+    AND test_name = %(test_name)s
+    AND input = %(inp)s
+
+    UNION
+
+    SELECT
+        anomaly,
+        confirmed,
+        input,
+        measurement_start_time,
+        probe_cc,
+        coalesce('') as scores,
+        test_name,
+        coalesce('') as tid
+
+    FROM measurement
+    JOIN report ON report.report_no = measurement.report_no
+    JOIN input ON input.input_no = measurement.input_no
+    WHERE measurement_start_time >= %(start_date)s
+    AND probe_cc = %(cc)s
+    AND test_name = %(test_name)s
+    AND input = %(inp)s
+
+    ORDER BY measurement_start_time
+    """
```

Comment:
I assume your end goal is to use this function to implement also the one above?

Reply:
Those are 2 different use cases: filtering by test_name/cc/input or getting everything.